### PR TITLE
MapTemplate plugin should be usable in localConfig (#5564)

### PR DESCRIPTION
**Pull request by AGMETEOR**
## Description
MapTemplate plugin should be usable in localConfig (#5564). This PR enables the MapTemplates plugin to load templates both from context object stored on geostore and localConfig i.e. 
```json
{
"name": "MapTemplates",
"cfg": {
"templates": [{"id": <template resource id>}]
}
}
```

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Bugfix
 - [ ] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue

**What is the current behavior?**

#5564 

**What is the new behavior?**
MapTemplates plugin can load templates from both context object and localConfig

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

**Comment by AGMETEOR** [comment]
> This PR fixes the bug but I think that we should do it a little better.
> 
> I think that saving the templates in the context root was an error that breaks the idependency of plugins.
> 
> 
> 
> For this reason I should keep the context templates plugin selector only for retro-compatibility with saved contexts and I propose a better solution that works for both contexts and main viewer in the same way.
> 
> 
> 
> - On plugin mount an action stores the templates list from the config into the maptemplates state (reducer). (e.g. state.maptemplates.allowedTemplates)
> 
> - The selector check that piece of state, and if missing it tests also the context state (with soft-failing, only for retro-compatiblity)
> 
> - Context should save the mapTemplates in the configuration of the single plugin, instead of the root of the context.
> 
> 
> 
> What do you think, can you try to apply this alternative solution? 
> 
> 
> 
> 

Thsnks @offtherailz , let me work on the alternative solution you have suggested.

**Comment by AGMETEOR** [comment]
> * On plugin mount an action stores the templates list from the config into the maptemplates state (reducer). (e.g. state.maptemplates.allowedTemplates)


@offtherailz for this point, I have noticed that localConfig MapTemplates configuration will not be loaded on mount when the plugin is opened from context.

**Comment by AGMETEOR** [comment]
> This PR fixes the bug but I think that we should do it a little better.
> I think that saving the templates in the context root was an error that breaks the idependency of plugins.
> 
> For this reason I should keep the context templates plugin selector only for retro-compatibility with saved contexts and I propose a better solution that works for both contexts and main viewer in the same way.
> 
> * On plugin mount an action stores the templates list from the config into the maptemplates state (reducer). (e.g. state.maptemplates.allowedTemplates)
> * The selector check that piece of state, and if missing it tests also the context state (with soft-failing, only for retro-compatiblity)
> * Context should save the mapTemplates in the configuration of the single plugin, instead of the root of the context.
> 
> What do you think, can you try to apply this alternative solution?

@offtherailz I changed the implementation. I also think for the last point *Context should save the mapTemplates in the configuration of the single plugin, instead of the root of the context.* , we can separate it out into a new issue like you had suggested.

File: web/client/plugins/MapTemplates.jsx [hunk 1]
```diff
@@ -32,16 +32,24 @@ import * as epics from '../epics/maptemplates';
  * @memberof plugins
  * @class
  * @name MapTemplates
+ * * @prop {object[]} cfg.allowedTemplates: A list of objects with map template ids used to load templates when not in context
```
Comment (offtherailz) [inline web/client/plugins/MapTemplates.jsx:35]:
```suggestion
 * @prop {object[]} cfg.allowedTemplates: A list of objects with map template ids used to load templates when not in context
```

Comment (AGMETEOR) [reply web/client/plugins/MapTemplates.jsx]: Fixed

**Comment by offtherailz** [comment]
please @AGMETEOR backport the PR to 2020.02.01

**Pull request merged**